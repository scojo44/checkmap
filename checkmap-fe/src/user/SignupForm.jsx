import React from 'react'
import {Link} from 'react-router-dom';
import {useForm} from 'react-hook-form'
import FormField from '../widgets/FormField';
import NewPasswordFields from '../widgets/NewPasswordFields';
// import './SignupForm.css'

export default function SignupForm({signup}) {
  const {register, handleSubmit, formState: {errors}} = useForm();

  return (
    <form className="SignupForm" onSubmit={handleSubmit(async newUser => signup(newUser))}>
      <h2>Sign Up</h2>
      <FormField name="username" label="Username" {...{register, errors}} validation={{required: 'Please choose a username'}}/>
      <NewPasswordFields {...{register, errors}}/>
      <FormField name="imageURL" label="Image URL (optional)" {...{register, errors}}/>
      <button type="submit">Sign up</button> or <Link to="/login">Log in</Link>
    </form>
  );
}
