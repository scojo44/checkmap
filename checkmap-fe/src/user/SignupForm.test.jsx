import {render, fireEvent, waitFor} from '@testing-library/react'
import {MemoryRouter} from 'react-router-dom'
import UserContext from '../UserContext';
import SignupForm from './SignupForm'

const mockSignup = vi.fn();
const context = {
  SITE_NAME: 'CheckMap'
};

describe('SignupForm Tests', () => {
  beforeEach(() => {
    mockSignup.mockClear();
  });

  it('Renders without crashing', () => {
    render(
      <MemoryRouter>
        <UserContext.Provider value={context}>
          <SignupForm/>
        </UserContext.Provider>
      </MemoryRouter>
    );
  });

  it('Matches snapshot', () => {
    const {asFragment} = render(
      <MemoryRouter>
        <UserContext.Provider value={context}>
          <SignupForm/>
        </UserContext.Provider>
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('Calls the passed function with the submitted values', async () => {
    const {getByText, getByLabelText} = render(
      <MemoryRouter>
        <UserContext.Provider value={context}>
          <SignupForm signup={mockSignup}/>
        </UserContext.Provider>
      </MemoryRouter>
    );

    const username = getByLabelText('Username:');
    const password = getByLabelText('Password:');
    const confirm = getByLabelText('Confirm Password:');
    const imageURL = getByLabelText('Image URL (optional):');
    const button = getByText('Sign up');

    fireEvent.change(username, {target: {value: 'testuser'}});
    fireEvent.change(password, {target: {value: 'secret123'}});
    fireEvent.change(confirm, {target: {value: 'secret123'}});
    fireEvent.change(imageURL, {target: {value: 'testuser.jpeg'}});
    fireEvent.click(button);

    await waitFor(() => expect(mockSignup).toHaveBeenCalled());
    await waitFor(() => expect(mockSignup).toHaveBeenCalledWith({
      username: 'testuser',
      password: 'secret123',
      confirm: 'secret123',
      imageURL: 'testuser.jpeg'
    }));
  });

  it('Does not call the passed function when confirm does not match password', async () => {
    const {getByText, getByLabelText} = render(
      <MemoryRouter>
        <UserContext.Provider value={context}>
          <SignupForm signup={mockSignup}/>
        </UserContext.Provider>
      </MemoryRouter>
    );

    const username = getByLabelText('Username:');
    const password = getByLabelText('Password:');
    const confirm = getByLabelText('Confirm Password:');
    const imageURL = getByLabelText('Image URL (optional):');
    const button = getByText('Sign up');

    fireEvent.change(username, {target: {value: 'testuser'}});
    fireEvent.change(password, {target: {value: 'secret123'}});
    fireEvent.change(confirm, {target: {value: 'secret456'}});
    fireEvent.change(imageURL, {target: {value: 'testuser.jpeg'}});
    fireEvent.click(button);

    await waitFor(() => expect(mockSignup).not.toHaveBeenCalled());
  });
});
