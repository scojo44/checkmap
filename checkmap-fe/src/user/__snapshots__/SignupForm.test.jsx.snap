// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignupForm Tests > Matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="SignupForm"
  >
    <h2>
      Sign Up
    </h2>
    <p
      class="FormField"
    >
      <label
        for="username"
      >
        Username: 
      </label>
      <input
        id="username"
        name="username"
      />
    </p>
    <p
      class="FormField"
    >
      <label
        for="password"
      >
        Password: 
      </label>
      <input
        id="password"
        name="password"
        type="password"
      />
    </p>
    <p
      class="FormField"
    >
      <label
        for="confirm"
      >
        Confirm Password: 
      </label>
      <input
        id="confirm"
        name="confirm"
        type="password"
      />
    </p>
    <p
      class="FormField"
    >
      <label
        for="imageURL"
      >
        Image URL (optional): 
      </label>
      <input
        id="imageURL"
        name="imageURL"
      />
    </p>
    <button
      type="submit"
    >
      Sign Up
    </button>
     or 
    <a
      href="/login"
    >
      Log in
    </a>
  </form>
</DocumentFragment>
`;
